rfc-capi-for-kvm-capk/README.md

```diff
@@ -0,0 +1,425 @@
+1. setting per controller finalizer (`operatorkit.giantswarm.io/<controller.Name>`)
+1. allow delete events to get [replayed](https://github.com/giantswarm/operatorkit/blob/master/docs/using_finalizers.md#control-flow) (by setting `finalizerskeptcontext.SetKept(ctx)`
+1. when we boot the server we also add two additional endpoints: `healtz` (service availability) and `version` (info about the go runtime)
+
+##### Alternatives in kubebuilder
+
+1. Metrics: `kubebuilder` already exposes a few metrics and we can always add new ones ([link](https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics))
+1. Events: manager has a `GetEventRecorderFor` method that can be passed to the controller to register events
+1. Finalizers: kubebuilder supports them too so we can implement all the stuff we need (pause, replay)
+1. Extra endpoints: manager has `AddHealthzCheck` and `AddReadyzCheck` methods that feel similar to the `healtz` endpoint. Not sure if we actively use the second one still.
+
+### Milestones
+
+#### MVP
+
+- Create Cluster
+- Delete Cluster
+- Scale Cluster
+- Upgrade Cluster
+- Pivoting 
+- Cofigure node size for Worker and Control Plane Nodes
+
+#### Production-readiness
+
+- Configure docker volume size
+- Configure kubelet volume size
+- Vertical pod autoscaler for operator pod
+- Configure DNS and NTP Servers
+- Docker bridge IP and network range 
+    This is needed by some customers due to conflicting IP ranges
+- Configure CIDRs for pod and service IPs 
+
+### Appendix A: Full Example Cluster
+
+#### Cluster
+
+```yaml
+kind: Cluster
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+metadata:
+  name: example
+spec:
+  controlPlaneEndpoint:
+    host: example.k8s.geckon.gridscale.kvm.gigantic.io
+    port: 6443
+  infrastructureRef:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: KVMCluster
+    name: example
+```
+
+#### KVMCluster
+
+```yaml
+kind: KVMCluser
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+metadata:
+  name: example
+spec:
+  controlPlaneEndpoint:
+    host: example.k8s.geckon.gridscale.kvm.gigantic.io
+    port: 6443
+```
+
+#### KVMachineTemplate (control plane)
+
+```yaml
+kind: KVMachineTemplate
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+metadata:
+  name: example-control-plane
+spec:
+  template:
+    spec:
+      resources:
+        cpu: 1
+        memory: 1Gi
+        disk: 20Gi
+        dockerVolumeSize: 5Gi
+```
+
+#### KubeadmControlPlane
+
+```yaml
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+kind: KubeadmControlPlane
+metadata:
+  name: example
+spec:
+  infrastructureTemplate:
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: KVMMachineTemplate
+    name: example-control-plane
+  kubeadmConfigSpec:
+    initConfiguration:
+      nodeRegistration:
+        name: '{{ ds.meta_data.hostname }}'
+        kubeletExtraArgs:
+          cloud-provider: kvm
```

Comment:
Maybe this is nit picking but I think cloud-provider: kvm will not work as this works only for cloud providers that are included in the upstream tree. So I think it should be empty or `external`.

Reply:
Good point. I'll change to one of these after reviewing docs.